src/compiler/checker.ts

```diff
@@ -13900,20 +13900,29 @@ namespace ts {
             return getTypeFromInference(inference);
         }
 
-        function* getUnmatchedProperties(source: Type, target: Type, requireOptionalProperties: boolean) {
+        function* getUnmatchedProperties(source: Type, target: Type, requireOptionalProperties: boolean, matchDiscriminantProperties: boolean) {
             const properties = target.flags & TypeFlags.Intersection ? getPropertiesOfUnionOrIntersectionType(<IntersectionType>target) : getPropertiesOfObjectType(target);
             for (const targetProp of properties) {
                 if (requireOptionalProperties || !(targetProp.flags & SymbolFlags.Optional)) {
                     const sourceProp = getPropertyOfType(source, targetProp.escapedName);
                     if (!sourceProp) {
                         yield targetProp;
                     }
+                    else if (matchDiscriminantProperties) {
+                        const targetType = getTypeOfSymbol(targetProp);
+                        if (targetType.flags & TypeFlags.Unit) {
+                            const sourceType = getTypeOfSymbol(sourceProp);
+                            if (!(sourceType.flags & TypeFlags.Any || getRegularTypeOfLiteralType(sourceType) === getRegularTypeOfLiteralType(targetType))) {

```

Comment:
I suggest to union all nested `if`s in one condition like this:
`isTargetTypeUnit && !isSourceTypeAny && !regularTypesEqual` to make intention more clear. But maybe it's not appropriate here.

Reply:
I'm going to keep it the way it is. I don't want to `getTypeOfSymbol` on the source more than once and I don't want to do it unless the target is a unit type.

Reply:
Is it correct that you don't want to do this because of execution performance reason? How about code performance and its readability? Does first one outweigh the latter?